### PR TITLE
[2/3] loopd: add macaroon authentication to the loop RPC server

### PR DESCRIPTION
**Pull request by guggero**
Fixes https://github.com/lightninglabs/loop/issues/194.

Depends on https://github.com/lightningnetwork/lnd/pull/4464 (and indirectly on https://github.com/lightningnetwork/lnd/pull/4463).

With this PR we first secure the loop daemon's RPC connection by encrypting the transmitted data with TLS.
Next we add macaroon authentication to the RPC to be able to restrict access to authorized users only.
With these two features enabled by default, the loop RPC port can be opened to the internet because unauthorized access is no longer possible.

By default only one macaroon is created for loop and stored in a file called `~/.loop/<network>/loop.macaroon`.~To give users the ability to fully customize their access with custom macaroons, the two functionalities `BakeMacaroon` and `ListPermissions` are copied from `lnd` and added to both the RPC and command line.~ --> Split this part off into its own PR: https://github.com/lightninglabs/loop/pull/285

### CODE REVIEW

**Comment by carlaKC** [comment]
> Depends on lightningnetwork/lnd#4464 (and indirectly on lightningnetwork/lnd#4463).

Just a procedural question about this dependency, is this PR now going to wait until we tag 0.12 to go in? Or are the lnd PRs going into `lnd v0.11.1`?

File: loopd/run.go
Comment (carlaKC) [inline loopd/run.go:200]: Check this error?

File: loopd/daemon.go
Comment (carlaKC) [inline loopd/daemon.go:388]: nit: `err2`-> `macaroonErr`/`shutdownErr`?

File: loopd/macaroons.go
Comment (carlaKC) [inline loopd/macaroons.go:24]: Nicee 🥇

File: loopd/macaroons.go
Comment (carlaKC) [inline loopd/macaroons.go:158]: Could this logging happen when we call `startMacaroonService`?

**Comment by guggero** [comment]
Yes, the dependent PRs are both candidates for `lnd v0.11.1`, so this PR has to wait for that to be released.
But it's a compile-time dependency only, not an RPC dependency.

File: loopd/daemon.go
Comment (carlaKC) [inline loopd/daemon.go:208]: move these to the previous commit where we check `!d.cfg.NoMacaroons`?

File: looprpc/client.proto
Comment (carlaKC) [inline looprpc/client.proto:728]: nit: newline

**Review by carlaKC — COMMENTED** [review]
Awesome to have thins functionality added to loop 🎉 
No major questions, since a lot of this is copied from lnd, just some nits from my first pass. Will do some testing for this as well. 

Perhaps also update the `README` since we're going to enable by default?

File: loopd/config.go
Comment (joostjager) [reply loopd/config.go:99]: Maybe we shouldn't even add the option `NoMacaroons`?

File: loopd/macaroons.go
Comment (joostjager) [inline loopd/macaroons.go:129]: Won't we end up in the same situation as with `lnd`? That people want to store the password on the box anyway, defeating the whole purpose?

**Review by joostjager — COMMENTED** [review]
Just quickly skimmed through the PR. Is it worth to split this PR before continuing into the review cycle? For example in three parts:
* Add TLS
* Add (single) macaroon
* Add macaroon baking of rpc

**Review by joostjager — COMMENTED** [review]
Tested and works. Was just wondering how I can verify that it is now actually using TLS?

File: loopd/config.go
Comment (guggero) [reply loopd/config.go:99]: Removed the flag. This makes it a bit harder to test when developing against the REST API for example as you can't just get some JSON in the browser. But if anyone needs this feature they can request it later on.

File: loopd/config.go
Comment (guggero) [reply loopd/config.go:287]: These are just two different ways to hold the same certificate. Once we're using it as a server certificate (for gRPC) and once as a client connect CA certificate (for REST). Because the REST proxy connects to gRPC and is a client.

But you are right, something's missing. The REST server also needs to serve over a TLS connection!

File: loopd/macaroons.go
Comment (guggero) [reply loopd/macaroons.go:34]: Multiple permissions here mean that all of them are required. So it's an _AND_ operation, not _OR_. The idea is that (with the bakery) users could create macaroons that are working for loop in or loop out exclusively. I don't have an explicit use case in mind but that is something I think could come in handy for users.

File: cmd/loop/main.go
Comment (guggero) [reply cmd/loop/main.go:409]: I'm not exactly sure what you mean by this comment. But I did split the commit into the TLS part and the macaroon part. We now touch the same code twice but the commits are separated by the functionality they touch.

File: loopd/macaroons.go
Comment (guggero) [reply loopd/macaroons.go:101]: I extended the TODO comment that if we want to add a password we should make it optional.

File: cmd/loop/main.go
Comment (guggero) [reply cmd/loop/main.go:201]: It doesn't look like it. The `loop.db` is still in the network sub directory when a loop dir is set.

File: loopd/config.go
Comment (guggero) [reply loopd/config.go]: I added the restriction as suggested. But I don't want to change the behavior of the other flags in this PR, that's why the TLS and macaroon files are stored in the same place as `loop.db` at the moment (unless explicitly specified differently).

**Comment by guggero** [comment]
I addressed all comments and also added a short section about TLS and macaroons to the README.

@joostjager you can verify if TLS is enabled by using `openssl s_client -connect <host>:<port>`. If you get presented a ceritifcate then TLS is turned on.

File: loopd/config.go
Comment (joostjager) [inline loopd/config.go:217]: Doesn't it need a config default similar to tls path?

File: loopd/macaroons.go
Comment (joostjager) [reply loopd/macaroons.go:34]: Ok, clear

File: loopd/macaroons.go
Comment (joostjager) [inline loopd/macaroons.go:102]: Does something still need to be done with this? I'd think that a default password in open source code provides a false sense of security?

File: cmd/loop/main.go
Comment (joostjager) [inline cmd/loop/main.go:397]: `getClientConn` has an error return value, so you could pass back an error from `readMacaroon`

File: cmd/loop/main.go
Comment (joostjager) [inline cmd/loop/main.go:189]: This logic seems to be slightly different from what happens from tls path. I'd think these are two file paths and should be handled exactly the same? Perhaps even via a helper function.

File: README.md
Comment (joostjager) [inline README.md:338]: "by default" sounds as if it can be turned off?

File: README.md
Comment (joostjager) [inline README.md:341]: loop dir (`--loopdir`) or loop data dir (`--datadir`)? It is still a bit confusing which file is sought where.

Comment (joostjager) [reply README.md:341]: Perhaps the command line help should clarify that `loopdir` and the other path flags are mutually exclusive?

File: cmd/loop/main.go
Comment (joostjager) [reply cmd/loop/main.go]: Discussed offline and want to keep in line with lnd.

File: loopd/run.go
Comment (guggero) [reply loopd/run.go:69]: Yes. Removed the `if` and changed the comment. The TLS config can still be `nil` but that will result in an error from the TLS library itself.

File: loopd/run.go
Comment (guggero) [reply loopd/run.go:223]: No, the update to the newest `lnd` version caused the linter to complain about this new, unhandled exception.

File: loopd/macaroons.go
Comment (guggero) [reply loopd/macaroons.go:54]: My thought was that it's probably most important to be able to separate them for the write operations. But to be consequent we should probably use the same pattern for the read operations as well. Fixed.

File: loopd/macaroons.go
Comment (guggero) [reply loopd/macaroons.go:102]: I changed the password to be an empty one and added some more information to the comment.

File: loopd/daemon.go
Comment (joostjager) [inline loopd/daemon.go]: If it crashes with `nil`, this isn't a working commit and possibly a sign that it should be squashed with another one. But non-blocking.

**Review by joostjager — APPROVED** [review]

**Comment by joostjager** [comment]
>But it's a compile-time dependency only, not an RPC dependency.

With grub, a comment like this isn't relevant anymore, because even for compile-time deps we need to wait for a release?

**Review by carlaKC — APPROVED** [review]
LGTM

File: loopd/daemon.go
Comment (guggero) [reply loopd/daemon.go:349]: Yeah, you're right. Fixed.

File: loopd/macaroons.go
Comment (guggero) [reply loopd/macaroons.go:175]: Ah, that comment was from when we still had the `NoMacaroons` flag. Removed the "if it was started" part.

**Comment by guggero** [comment]
> With grub, a comment like this isn't relevant anymore, because even for compile-time deps we need to wait for a release?

We have to wait for a release (of lnd and loop that is) in LiT. But we can merge this PR as soon as the dependent PRs are merged to master. That's the main difference between an RPC and compile time dep IMO.

**Comment by joostjager** [comment]
But won't your compile-time required code disappear when loop is embedded in grub and the dependency is overriden with the last release 0.11.0?

**Comment by Kixunil** [comment]
Awesome! @guggero do you have a LN donation page? Would like to send you a tip for doing this. :)

**Comment by guggero** [comment]
Thanks for the feedback, I'm happy to hear this is useful! No need for a tip though, I am getting paid to work on these projects ;-)

**Pull request merged**
